Code to review (#1)
carlphilipp/compliance-logging: README.md

<p align="center"><img src="docs/images/slalom.png" align="center" width="300" hspace="10" vspace="10"></p>
<h6 align="center">presents</h6>
<h3 align="center"><b>Compliance Logging</b></h3>
<h4 align="center">A library that allows developers to mask sensitive information when logging.</h4>

<p align="center">
<a href="LICENSE" alt="Project License"><img src="https://img.shields.io/apm/l/vim-mode?style=for-the-badge"></a>
</p>

### Description

> This library has been built to help developers masking PII, PCI and PHI when logging with their favorite framework.<br />

Reviewer: ``
```suggestion
This library has been built to help developers mask PII, PCI and PHI when logging using with their favorite framework.<br />
```

Author: Applied all your comments. Thank you!

For instance, to log a JSON String that contains some sensitive information, the result would be something like:

```log
12:44:09 [main] INFO  MyClass - {"login":"john","password":"***********"}
```

At the moment, the library supports two types of objects that can be scanned to mask sensitive information: JSON and Lombok.<br />
In order to achieve that, the developer needs to provide the framework with a custom marker. <br />
For example:
```log
log.info("{}", "{\"login\":\"john\",\"password":\"mypassword\"});
```

> Would not mask anything, while if a mask is used like below:

Reviewer: ```suggestion
Would not mask anything; however, if a mask is used like below:
```

```log
log.info(MaskType.JSON, "{}", "{\"login\":\"john\",\"password":\"mypassword\"});
```

> It would mask the password (or any other fields listed as sensitive).

Reviewer: ```suggestion
It would mask the password (or any other fields that have been configured as sensitive).
```

Checkout our [examples](examples) in our repository.

### Usage with Log4j2
Import the latest version of the library in your project:<br />
Maven:
```xml
<dependency>
    <groupId>com.slalom</groupId>
    <artifactId>compliance-logging-log4j2</artifactId>
    <version>{LATEST_VERSION}</version>
</dependency>
```
Gradle (Kotlin DSL):
```kotlin
dependencies {
    implementation(group = "com.slalom", name = "compliance-logging-log4j2", version = "{LATEST_VERSION}")
}
```

> Then three things needs to be setup in your Log4j2 configuration:

Reviewer: ```suggestion
Then three things need to be setup in your Log4j2 configuration:
```

1. Package scanning (to scan our plugin).
```xml
<Configuration tatus="WARN" packages="com.slalom.logging">
  ...
</Configuration>
```

> 2. Add a property that represents the list of fields that need to be masked, the separator being a coma.

Reviewer: ```suggestion
2. Add a property that represents the list of fields that need to be masked, the separator being a comma.
```

```
Full example:
```xml
<?xml version="1.0" encoding="UTF-8"?>
<Configuration tatus="WARN" packages="com.slalom.logging">
    <Properties>
        <Property name="COMPLIANCE_FIELDS">password,ssn</Property>
        <Property name="CONSOLE_LOG_PATTERN">%d{HH:mm:ss} [%t] %-5level %logger{36} - %mm{${COMPLIANCE_FIELDS}}%n
        </Property>
    </Properties>
    <appenders>
        <Console name="Console" target="SYSTEM_OUT">
            <PatternLayout pattern="${sys:CONSOLE_LOG_PATTERN}"/>
        </Console>
    </appenders>
    <loggers>
        <root level="info">
            <appender-ref ref="Console"/>
        </root>
    </loggers>
</Configuration>
```
[Simple example](examples/simple-log4j2-example) - [Spring Boot example](examples/spring-boot-log4j2-example)

### Usage with Logback
Import the latest version of the library in your project:<br />
Maven:
```xml
<dependency>
    <groupId>com.slalom</groupId>
    <artifactId>compliance-logging-logback</artifactId>
    <version>{LATEST_VERSION}</version>
</dependency>
```
Gradle (Kotlin DSL):
```kotlin
dependencies {
    implementation(group = "com.slalom", name = "compliance-logging-logback", version = "{LATEST_VERSION}")
}
```

> Then two things needs to be setup in your Logback configuration:

Reviewer: ```suggestion
Then two things need to be setup in your Logback configuration:
```

1. Add our Layout (`com.slalom.logging.compliance.logback.PatternMaskingLayout`).

> 2. Provide the fields that needs to be masked, the separator being a coma.

Reviewer: ```suggestion
2. Provide the fields that need to be masked as a comma separated list.
```

        </Property>
    </Properties>
    <appenders>
        <Console name="Console" target="SYSTEM_OUT">
            <PatternLayout pattern="${sys:CONSOLE_LOG_PATTERN}"/>
        </Console>
    </appenders>
    <loggers>
        <root level="info">
            <appender-ref ref="Console"/>
        </root>
    </loggers>
</Configuration>
```
[Simple example](examples/simple-log4j2-example) - [Spring Boot example](examples/spring-boot-log4j2-example)

### Usage with Logback
Import the latest version of the library in your project:<br />
Maven:
```xml
<dependency>
    <groupId>com.slalom</groupId>
    <artifactId>compliance-logging-logback</artifactId>
    <version>{LATEST_VERSION}</version>
</dependency>
```
Gradle (Kotlin DSL):
```kotlin
dependencies {
    implementation(group = "com.slalom", name = "compliance-logging-logback", version = "{LATEST_VERSION}")
}
```
Then two things need to be setup in your Logback configuration:

1. Add our Layout (`com.slalom.logging.compliance.logback.PatternMaskingLayout`).
2. Provide the fields that need to be masked, the fields should be given as a comma separated list.

> Here is a full example:

Reviewer 2: Did we have a yaml based example as well?

Author: I did not, but I added one in the spring-boot-log4j2 example. I'm not too concerned about that because this is more a log4j2 thing. We could for example use a `.properties` or even a `.json` file aswell
